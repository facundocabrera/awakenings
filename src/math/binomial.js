// https://www.khanacademy.org/math/statistics-probability/random-variables-stats-library/binomial-random-variables/v/probability-of-making-2-shots-in-6-attempts

import { pow, combinations } from "mathjs";
import range from "lodash/range";
import { cumulative } from "./stats";

/**
 * Binomial Mass.
 *
 * @param {Number} n independent trials
 * @param {Number} k number of times for a specific outcome within n trials
 * @param {Number} p successes probability for a single trial
 *
 * @returns Number between [0, 1]
 */
const mass = (n, k, p) => combinations(n, k) * pow(p, k) * pow(1 - p, n - k);

/**
 * Binomial distribution.
 *
 * Comienzo desde 0 exitos, hasta evaluar n exitos en los n intentos.
 *
 * @param {Number} n independent trials
 * @param {Number} p successes probability for a single trial
 */
const distribution = (n, p) => range(0, n).map((s) => mass(n, s, p));

export { mass, distribution, cumulative };
