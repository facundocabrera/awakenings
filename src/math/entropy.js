import log from "./log";

/**
 * Self-information of an event.
 * 
 * @param {Number} probability of the event
 * @param {Number} base unit of self-information
 */
const I = (probability, base) => -1 * log(probability, base);

/**
 * Entropy.
 * 
 * Se puede interpretar como el valor esperado para un conjunto de 
 * self-informative events. Basicamente es la mean usando la funcion I.
 * 
 * @param {Array} distribution probabilities for each different event
 * @param {Number} base base to be used in the log calculation
 * 
 * @returns entropy Number between [0, 1]
 */
const H = (distribution, base = 10) => distribution.reduce(
  (h, p) => h -= p * log(p, base), 0
);

export {
  I,
  H
};