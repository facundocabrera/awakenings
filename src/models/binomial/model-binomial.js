import { sum } from "lodash";

const { combinations } = require("mathjs");

const { PI } = Math;

const build = (n) => {
  const comb = [];

  for (let i = 0; i <= n; i++) comb.push(Math.round(combinations(n, i)));

  return comb;
};

const prob = (counters) => {
  const total = sum(counters);

  return counters.map((counter) => counter / total);
};

const lengths = (counters) =>
  counters.map((counters) => String(counters).length);

const angles = (prob) => prob.map((p) => p * 2 * PI);

export { build, prob, lengths, angles };
