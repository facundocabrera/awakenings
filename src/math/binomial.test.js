// https://en.wikipedia.org/wiki/Binomial_distribution

import { cumulative, distribution } from "./binomial";
import { H } from "./entropy";

test("math/binomial", () => {
  const n = 11;
  const p = 0.5;
  
  const d = distribution(n, p);
  const accum = cumulative(d);
  const { length: l } = accum;

  // la distribucion accumulada me da aproximadamente el 50% en la mitad
  expect(
    Math.abs(
        accum[(l % 2 === 0) ? l / 2 - 1 : (l - 1) / 2 + 1 ] - 0.5
      ) < 0.01
    ).toBe(true);
});

test("math/binomial/entropy", () => {
  const n = 16;
  const p = 0.5;
  const d = distribution(n, p);
  const ent = H(d);

  expect(ent).toMatchSnapshot();
});
