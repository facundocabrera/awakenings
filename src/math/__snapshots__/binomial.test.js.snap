// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`math/binomial/cumulative-for-6-trials 1`] = `
Array [
  0.3348979766803842,
  0.401877572016461,
  0.20093878600823048,
  0.053583676268861444,
  0.008037551440329218,
  0.0006430041152263372,
]
`;

exports[`math/binomial/cumulative-for-6-trials 2`] = `
Array [
  0.3348979766803842,
  0.7367755486968453,
  0.9377143347050757,
  0.9912980109739371,
  0.9993355624142664,
  1,
]
`;

exports[`math/binomial/entropy 1`] = `0.9169558133643613`;
